analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ConditionEvaluatesToConstantBase.cs

```diff
@@ -75,16 +75,19 @@ public override ProgramState ConditionEvaluated(SymbolicContext context)
     private bool IsIgnored(ProgramState state, IOperation operation) =>
         operation.Kind is OperationKindEx.Literal
         || IsVarPattern(operation)
-        || IsDiscardPattern(operation)
+        || IsDiscardPattern(state, operation)
         || operation.TrackedSymbol(state) is IFieldSymbol { IsConst: true }
         || operation.Syntax.Ancestors().Any(x => IsInsideUsingDeclaration(x) || IsLockStatement(x));
 
     private static bool IsVarPattern(IOperation operation) =>
         operation.AsIsPattern()?.Pattern.WrappedOperation.AsDeclarationPattern()?.MatchesNull is true;
 
-    private static bool IsDiscardPattern(IOperation operation) =>
-        operation.AsIsPattern() is { } pattern
-        && pattern.Pattern.WrappedOperation.Kind is OperationKindEx.DiscardPattern;
+    private static bool IsDiscardPattern(ProgramState state, IOperation operation) =>
+        operation.Kind is OperationKindEx.DiscardPattern
+        || (operation.AsIsPattern() is { } isPattern
+            && state.ResolveCaptureAndUnwrapConversion(isPattern.Pattern.WrappedOperation) is var subPattern
+            && (subPattern.Kind is OperationKindEx.DiscardPattern
+                || (subPattern.AsRecursivePattern() is {} recursivePattern && recursivePattern.DeconstructionSubpatterns.Any(x => IsDiscardPattern(state, x)))));
```

Comment:
This also needs to look for `var` patterns; See the method above this one. The two methods probably need to be merged.

Reply:
I tried that, but if I merge the two it will create some FNs:

```cs
public void ParenthesizedVariableDesignation(object arg)
{
    switch ((arg, true))
    {
        case var (o, b) when b:         // Will become FN
            if (b) { }                  // Will become FN
            break;

        ...
    }
}
```

Getting the variable declarations correct for this rule doesn't seem trivial, I'd rather leave it to a separate PR.